packages/core/echo/echo-db/src/database.ts

```diff
@@ -129,27 +163,115 @@ export class EchoDatabaseImpl extends Resource implements EchoDatabase {
+  ): Promise<Array<EchoReactiveObject<any> | undefined>> {
+    const result = new Array(objectIds.length);
+    const objectsToLoad: Array<{ id: string; resultIndex: number }> = [];
+    for (let i = 0; i < objectIds.length; i++) {
+      const objectId = objectIds[i];
+      const object = this.getObjectById(objectId);
+      if (this._automerge._objects.get(objectId)?.isDeleted()) {
+        result[i] = undefined;
+      } else if (object != null) {
+        result[i] = object;
+      } else {
+        objectsToLoad.push({ id: objectId, resultIndex: i });
+      }
+    }
+    if (objectsToLoad.length === 0) {
+      return result;
+    }
+    const idsToLoad = objectsToLoad.map((v) => v.id);
+    this.automerge._automergeDocLoader.loadObjectDocument(idsToLoad);
+
+    return new Promise((resolve, reject) => {
+      let unsubscribe: UnsubscribeCallback | null = null;
+      let inactivityTimeoutTimer: any | undefined;
+      const scheduleInactivityTimeout = () => {
+        inactivityTimeoutTimer = setTimeout(() => {
+          unsubscribe?.();
+          reject(new TimeoutError(inactivityTimeout));
+        }, inactivityTimeout);
+      };
+      unsubscribe = this._automerge._updateEvent.on(({ itemsUpdated }) => {
+        const updatedIds = itemsUpdated.map((v) => v.id);
+        for (let i = objectsToLoad.length - 1; i >= 0; i--) {
+          const objectToLoad = objectsToLoad[i];
+          if (updatedIds.includes(objectToLoad.id)) {
+            clearTimeout(inactivityTimeoutTimer);
+            result[objectToLoad.resultIndex] = this._automerge._objects.get(objectToLoad.id)?.isDeleted()
+              ? undefined
+              : this.getObjectById(objectToLoad.id)!;
+            objectsToLoad.splice(i, 1);
+            scheduleInactivityTimeout();
+          }
+        }
+        if (objectsToLoad.length === 0) {
+          clearTimeout(inactivityTimeoutTimer);
+          unsubscribe?.();
+          resolve(result);
+        }
+      });
+      scheduleInactivityTimeout();
+    });
   }
 
   add<T extends ReactiveObject<any>>(obj: T): EchoReactiveObject<{ [K in keyof T]: T[K] }> {
-    if (isEchoObject(obj)) {
-      this._automerge.add(obj);
-      return obj as any;
-    } else {
+    let echoObject = obj;
+    if (!isEchoObject(echoObject)) {
       const schema = getSchema(obj);
 
       if (schema != null) {
         if (!this.schemaRegistry.isRegistered(schema) && !this.graph.runtimeSchemaRegistry.hasSchema(schema)) {
           throw createSchemaNotRegisteredError(schema);
         }
       }
-
-      const echoObj = createEchoObject(obj);
-      this._automerge.add(echoObj);
-      return echoObj as any;
+      echoObject = createEchoObject(obj);
     }
+    invariant(isEchoObject(echoObject));
+    this._rootProxies.set(getAutomergeObjectCore(echoObject), echoObject);
+    this._automerge.add(echoObject);
+    EchoReactiveHandler.instance.drainLinkCache(echoObject as any);
```

Comment:
drainLinkCache before adding it to low level db

Reply:
some tests fail if I do this, so added TODO for future